query/src/main/scala/filodb/query/exec/rangefn/AggrOverTimeFunctions.scala

```diff
@@ -669,3 +670,114 @@ class QuantileOverTimeChunkedFunctionL(funcParams: Seq[Any])
   }
 }
 
+abstract class HoltWintersChunkedFunction(funcParams: Seq[Any],
+                                          var b0: Double = Double.NaN,
+                                          var s0: Double = Double.NaN,
+                                          var nextvalue: Double = Double.NaN)
+  extends ChunkedRangeFunction[TransientRow] {
+
+  override final def reset(): Unit = { s0 = Double.NaN
+                                       b0 = Double.NaN
+                                       nextvalue = Double.NaN }
+
+  def parseParameters(funcParams: Seq[Any]): (Double, Double) = {
+    require(funcParams.size == 2, "Holt winters needs 2 parameters")
+    require(funcParams.head.isInstanceOf[Number], "sf parameter must be a number")
+    require(funcParams(1).isInstanceOf[Number], "tf parameter must be a number")
+    val sf = funcParams.head.asInstanceOf[Number].doubleValue()
+    val tf = funcParams(1).asInstanceOf[Number].doubleValue()
+    require(sf >= 0 & sf <= 1, "Sf should be in between 0 and 1")
+    require(tf >= 0 & tf <= 1, "tf should be in between 0 and 1")
+    (sf, tf)
+  }
+
+  final def apply(endTimestamp: Long, sampleToEmit: TransientRow): Unit = {
+    sampleToEmit.setValues(endTimestamp, s0)
+  }
+}
+
+class HoltWintersChunkedFunctionD(funcParams: Seq[Any]) extends HoltWintersChunkedFunction(funcParams)
+  with ChunkedDoubleRangeFunction {
+
+  // Returns the first non-Nan value encountered
+  def getNextValue(startRowNum: Int, endRowNum: Int, it: DoubleIterator): (Double, Int) = {
+    var res = Double.NaN
+    var currRowNum = startRowNum
+    while (currRowNum <= endRowNum && JLDouble.isNaN(res)) {
+      val nextvalue = it.next
+      // There are many possible values of NaN.  Use a function to ignore them reliably.
+      if (!JLDouble.isNaN(nextvalue)) {
+        res = nextvalue
+      }
+      currRowNum += 1
+    }
+    (res, currRowNum)
+  }
+
+  final def addTimeDoubleChunks(doubleVect: BinaryVector.BinaryVectorPtr,
+                                doubleReader: bv.DoubleVectorDataReader,
+                                startRowNum: Int,
+                                endRowNum: Int): Unit = {
+    val (sf, tf) = parseParameters(funcParams)
```

Comment:
I would move `parseParameters` out of this function.  `addTimeDoubleChunks` could be called repeatedly for many chunks, and you don't want to parse it multiple times since the params are the same.